diff --git a/build/module/lib/electrum-ws.js b/build/module/lib/electrum-ws.js
index 5d281a2f531ca95ea72dc565a8e6f807983c3e02..45936f20bda76750b9aac2255b8287d03ad65015 100644
--- a/build/module/lib/electrum-ws.js
+++ b/build/module/lib/electrum-ws.js
@@ -186,7 +186,7 @@ export class ElectrumWS extends Observable {
         // Handle potential multi-line frames
         const raw = typeof msg.data === 'string' ? msg.data : bytesToString(msg.data);
         // eslint-disable-next-line no-control-regex
-        const regExpNewLineOrBlank = new RegExp('\r|\n| ', 'g');
+        const regExpNewLineOrBlank = new RegExp('\r|\n', 'g');
         const lines = raw
             .split(regExpNewLineOrBlank)
             .filter((line) => line.length > 0);